mono/utils/mono-threads.c

```diff
@@ -1117,6 +1136,21 @@ mono_thread_info_begin_resume (MonoThreadInfo *info)
 	return mono_thread_info_core_resume (info);
 }
 
+gboolean
+mono_thread_info_begin_pulse_resume_and_request_suspension (MonoThreadInfo *info)
+{
+	/* For two-phase suspend, we want to atomically resume the thread and
+	 * request that it try to cooperatively suspend again.  Specifically,
+	 * we really don't want it to transition from GC Safe to GC Unsafe
+	 * because we then it could (in GC Unsafe) try to take a lock that's
+	 * held by another preemptively-suspended thread, essentially
+	 * recreating the same problem that two-phase suspend intends to
+	 * fix. */
+	if (!mono_threads_is_hybrid_suspension_enabled ())
```

Comment:
Shouldn't this be `if (! mono_threads_is_hybrid_suspension_enabled() && ! mono_threads_is_cooperative_suspension_enabled())`?

Reply:
It should be a (non-existent) `mono_threads_is_multiphase_stw_enabled()`: we only need to pulse if STW requires two-phase suspension.  If there's only a single phase, resuming is fine.